### ISSUE TITLE
Should UseAutofacMiddleware dispose container on OWIN shutdown?

### ISSUE THREAD

**Issue by tillig** [autofac/Autofac.WebApi.Owin#1]
_From @chriseldredge on August 21, 2014 20:48_

I currently have this method in my OWIN Startup class:

``` c#
protected virtual void RegisterShutdownCallback(IAppBuilder app, IContainer container)
{
    var context = new OwinContext(app.Properties);
    var token = context.Get<CancellationToken>("host.OnAppDisposing");

    if (token != CancellationToken.None)
    {
        token.Register(container.Dispose);
    }
}
```

My question is should UseAutofacMiddleware do this for me?

On the one hand, the `IContainer` instance is supplied externally, so there may be cases where the container should not be disposed simply because the OWIN app is shutting down.

On the other hand, the default behavior means that containers won't be disposed properly when a self-hosted app or IIS app shuts down. This can cause problems for containers that have services that need to clean up before the process exits.

_Copied from original issue: autofac/Autofac#565_

**Comment by tillig** [comment]
It would probably be a decent thing to offer if people want to use it, but I don't know that we'd want to do it by default. Our stand thus far on things like disposing of lifetime scopes and containers has been "if you create it, you dispose it." But I think adding a little convenience method people can optionally use wouldn't hurt any.

**Comment by tillig** [comment]
_From @chriseldredge on October 28, 2014 17:32_

I agree "if you create it, you dispose it", but I guess I was thinking it would be nice to not have to reach into Owin properties and register my own shutdown listener in all my apps. I guess this is just the lay of the land.

**Comment by tillig** [comment]
For now, yeah, you'll have to register your own, but what I'm proposing is that we wrap up that ugliness for you so you can optionally do something like `app.DisposeContainerOnShutdown(container)` and that we'd include that extension for you. We'll document it and if people want to use it they can; or if not, they don't have to.

I'm hoping that providing the method but not just auto-disposing will be a reasonable compromise to allow both flexibility and ease of use.

**Comment by tillig** [comment]
_From @chriseldredge on October 29, 2014 14:58_

:+1:

**Comment by tillig** [comment]
_From @angelyordanov on November 28, 2014 17:1_

@chriseldredge your method looks great, why don't you make it a pull request. I guess it should be in `Core/Source/Autofac.Integration.Owin/AutofacAppBuilderExtensions.cs`. I can do it too, but its your code :).

**Comment by tillig** [comment]
Implemented with PR #2.

**Comment by tillig** [comment]
This change has been pushed to NuGet as [v3.3.0](https://www.nuget.org/packages/Autofac.WebApi2.Owin/3.3.0).
